enclave/core/sgx/exception.c

```diff
@@ -456,29 +463,37 @@ void oe_real_exception_dispatcher(oe_context_t* oe_context)
         oe_abort();
         return;
     }
-    else if (handler_ret == OE_EXCEPTION_ABORT_EXECUTION)
+
+    if (is_segfault)
+        OE_TRACE_ERROR("Segmentation fault");
+    else
+        OE_TRACE_ERROR("Unhandled exception");
+
+    void* buffer[OE_BACKTRACE_MAX];
+    oe_result_t r = OE_UNEXPECTED;
+    int size;
+
+    /* Get the backtrace from the previous frame where the exception occurs */
+    if ((size = _backtrace_skip_current_frame(buffer, OE_BACKTRACE_MAX)) > 0)
```

Comment:
Do we need this or can we directly pass the rbp from the exception record to get backtrace?

Reply:
There are two options:
1. Use `oe_backtrace_impl(__builtin_frame_address(0), buffer, OE_BACKTRACE_MAX)` directly so that we can bypass this extra function call.
   ```c
   if ((size = oe_backtrace_impl(__builtin_frame_address(0), buffer, OE_BACKTRACE_MAX)) > 0)
   {
       ...
   }
   ```

   With this option, we will only print the correct backtrace in the debug mode where the stack is stitched:
https://github.com/openenclave/openenclave/blob/4c431be5e3502b8f639ab1a107f8cb29d203b5bc/enclave/core/sgx/exception.c#L390

2. Construct the fake frame based on rbp and rip from the execution record.
   ```c
    uint64_t debug_frame[2];
    debug_frame[0] = oe_exception_record.context->rbp;
    debug_frame[1] = oe_exception_record.context->rip;

   if ((size = oe_backtrace_impl((void**)debug_frame, buffer, OE_BACKTRACE_MAX)) > 0)
   {
       ...
   }
   ```

   This option should work regardless the stack is stitched or not (as long as frame is preserved).


Which one do you think is better?

Reply:
I use option 2 for now.